### PR TITLE
[SYCLomatic] Add diagnostic when synchronizing thread blocks has side effects

### PR DESCRIPTION
**Pull request by yihwang-nv**
Signed-off-by: Wang, Yihan [yihan.wang@intel.com](mailto:yihan.wang@intel.com)

### CODE REVIEW

File: clang/lib/DPCT/ASTTraversal.cpp [hunk 1]
```diff
@@ -12046,6 +12197,33 @@ void SyncThreadsRule::runRule(const MatchFinder::MatchResult &Result) {
   }
 }
 
+void SyncThreadsRule::checkCallSyncthreadHasSideEffects(const CallExpr *CE) {
```
Comment (yihwang-nv) [inline clang/lib/DPCT/ASTTraversal.cpp]: FIXME: Cache the result to avoid redundant search and diagnostic.

Comment (yihwang-nv) [reply clang/lib/DPCT/ASTTraversal.cpp]: done

File: clang/lib/DPCT/ASTTraversal.cpp [hunk 2]
```diff
@@ -11894,6 +11894,202 @@ void CooperativeGroupsFunctionRule::runRule(
...
+    return const_ref_range(SideEffectCallExprs.begin(),
+                           SideEffectCallExprs.end());
+  }
+};
+
+} // namespace
+
+//
+void SyncThreadsRule::checkCallSyncthreadHasSideEffects(const CallExpr *CE) {
+  if (!CE)
```
Comment (tangjj11) [inline clang/lib/DPCT/ASTTraversal.cpp]: CE must not null, needn't this if, and can use reference parameter.

File: clang/lib/DPCT/ASTTraversal.cpp [hunk 3]
```diff
@@ -11894,6 +11894,202 @@ void CooperativeGroupsFunctionRule::runRule(
...
+      llvm::iterator_range<decltype(SideEffectCallExprs)::const_iterator>;
+  const_ref_range calls() const {
+    return const_ref_range(SideEffectCallExprs.begin(),
+                           SideEffectCallExprs.end());
+  }
+};
+
+} // namespace
+
+//
```
Comment (tangjj11) [inline clang/lib/DPCT/ASTTraversal.cpp]: need comment.

File: clang/lib/DPCT/ASTTraversal.cpp (hunk 2 above)
Comment (yihwang-nv) [reply clang/lib/DPCT/ASTTraversal.cpp]: done

File: clang/lib/DPCT/ASTTraversal.cpp (hunk 3 above)
Comment (yihwang-nv) [reply clang/lib/DPCT/ASTTraversal.cpp]: done

File: clang/test/dpct/syncthreads.cu [hunk 4]
```diff
@@ -78,10 +81,12 @@ __global__ void test2(float *ptr1, float *ptr2) {
   for (;;) {
     ptr1[aaa];
     // CHECK:item_ct1.barrier(sycl::access::fence_space::local_space);
+    // CHECK:DPCT1112:{{.*}}: Synchronizing thread blocks inside a conditional branch statement may lead to undefined behavior
```
Comment (zhimingwang36) [inline clang/test/dpct/syncthreads.cu]: Line84 is not necessary, in L81 foo, there is no condition at all.

File: clang/test/dpct/syncthreads.cu [hunk 5]
```diff
@@ -78,10 +81,12 @@ __global__ void test2(float *ptr1, float *ptr2) {
...
     __syncthreads();
 #pragma unroll
     for (;;) {}
     // CHECK:item_ct1.barrier(sycl::access::fence_space::local_space);
+    // CHECK:DPCT1112:{{.*}}: Synchronizing thread blocks inside a conditional branch statement may lead to undefined behavior
```
Comment (zhimingwang36) [inline clang/test/dpct/syncthreads.cu]: same as above.

File: clang/test/dpct/syncthreads_in_cond.cu [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+  // while
+  while (a[tid])
+    // CHECK: DPCT1112:{{.*}}: Synchronizing thread blocks inside a conditional branch statement may lead to undefined behavior
+    __syncthreads();
+
+  // early return
+  if (tid < 32)
+    return;
+
+  // CHECK: DPCT1112:{{.*}}: Synchronizing thread blocks inside a conditional branch statement may lead to undefined behavior
```
Comment (zhimingwang36) [inline clang/test/dpct/syncthreads_in_cond.cu]: let's not emit Line62. instead emit in the call stub of test1.

File: clang/test/dpct/syncthreads_in_cond.cu [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+__device__ void test1(uint64_t *a) {
+  unsigned tid =
+      ((blockIdx.x + (blockIdx.y * gridDim.x)) * (blockDim.x * blockDim.y)) +
+      (threadIdx.x + (threadIdx.y * blockDim.x));
+
+  // check-achor-begin-1
+  __syncthreads();
+  // check-achor-end-1
+  // CHECK: check-achor-begin-1
+  // CHECK-NOT: DPCT1112:{{.*}}: Synchronizing thread blocks inside a conditional branch statement may lead to undefined behavior
```
Comment (zhimingwang36) [inline clang/test/dpct/syncthreads_in_cond.cu]: let's not emit L17, instead emit in call stub of test1()

File: clang/test/dpct/syncthreads_in_cond.cu [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+
+  // check-achor-begin-2
+  test2(a);
+  // check-achor-end-2
+  // CHECK: check-achor-begin-2
+  // CHECK-NOT: DPCT1112:{{.*}}: Synchronizing thread blocks inside a conditional branch statement may lead to undefined behavior
+  // CHECK: check-achor-end-2
+
+  if (threadIdx.x > 10)
+    // CHECK: DPCT1112:{{.*}}: Synchronizing thread blocks inside a conditional branch statement may lead to undefined behavior
```
Comment (zhimingwang36) [inline clang/test/dpct/syncthreads_in_cond.cu]:
```suggestion
    // CHECK: DPCT1112:{{.*}}: The synchronizing thread blocks inside a conditional branch statement in test2() may lead to undefined behavior
```

File: clang/test/dpct/syncthreads_in_cond.cu [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+    test2(a);
+
+  if (threadIdx.x < 5)
+    return;
+  // CHECK: DPCT1112:{{.*}}: Synchronizing thread blocks inside a conditional branch statement may lead to undefined behavior
+  test2(a);
+
+  for (int i = 0; i < 10; ++i) {
+    // CHECK: DPCT1112:{{.*}}: Synchronizing thread blocks inside a conditional branch statement may lead to undefined behavior
```
Comment (zhimingwang36) [inline clang/test/dpct/syncthreads_in_cond.cu]: Please remove the L98,  as the condition is quite clear, each thread will come here.

File: clang/test/dpct/syncthreads.cu (hunk 4 above)
Comment (yihwang-nv) [reply clang/test/dpct/syncthreads.cu]: done

File: clang/test/dpct/syncthreads.cu (hunk 5 above)
Comment (yihwang-nv) [reply clang/test/dpct/syncthreads.cu]: done

File: clang/test/dpct/syncthreads_in_cond.cu (hunk 6 above)
Comment (yihwang-nv) [reply clang/test/dpct/syncthreads_in_cond.cu]: done

File: clang/test/dpct/syncthreads_in_cond.cu (hunk 8 above)
Comment (yihwang-nv) [reply clang/test/dpct/syncthreads_in_cond.cu]: done

File: clang/test/dpct/syncthreads_in_cond.cu (hunk 9 above)
Comment (yihwang-nv) [reply clang/test/dpct/syncthreads_in_cond.cu]: done

File: clang/test/dpct/syncthreads_in_cond.cu (hunk 7 above)
Comment (yihwang-nv) [reply clang/test/dpct/syncthreads_in_cond.cu]: done

File: clang/lib/DPCT/ASTTraversal.cpp [hunk 10]
```diff
@@ -11994,6 +12004,17 @@ void SyncThreadsRule::runRule(const MatchFinder::MatchResult &Result) {
       CE->getDirectCallee()->getNameInfo().getName().getAsString();
   if (FuncName == "__syncthreads") {
     DpctGlobalInfo::registerNDItemUser(CE);
+    const FunctionDecl *FD = nullptr;
+    if ((FD = getAssistNodeAsType<FunctionDecl>(Result, "FuncDecl")) ||
+        (FD = getAssistNodeAsType<FunctionDecl>(Result, "FuncDeclUsed"))) {
```
Comment (yihwang-nv) [inline clang/lib/DPCT/ASTTraversal.cpp]: **FIXME:** The return type of `__syncthreads()` is `void`, we don't need to check `FuncDeclUsed`.

**Review by AndyCHHuang — APPROVED** [review]
LGTM

File: clang/lib/DPCT/ASTTraversal.cpp (hunk 10 above)
Comment (yihwang-nv) [reply clang/lib/DPCT/ASTTraversal.cpp]: done

File: clang/lib/DPCT/Diagnostics.inc [hunk 11]
```diff
@@ -243,3 +243,6 @@ DEF_WARNING(ONE_DIMENSION_KERNEL_BARRIER, 1113, "Consider replacing sycl::nd_ite
 DEF_COMMENT(ONE_DIMENSION_KERNEL_BARRIER, 1113, "Consider replacing sycl::nd_item::barrier(sycl::access::fence_space::local_space) with sycl::nd_item::barrier() if function \"{0}\" is called in a multidimensional kernel.")
 DEF_WARNING(WAIT_REMOVE, 1114, "cudaMemcpy is migrated to asynchronization memcpy, assuming in original code the source host memory is pageable memory. If this assumption doesn't hold, please append '.wait()' after memcpy to synchronize.")
 DEF_COMMENT(WAIT_REMOVE, 1114, "cudaMemcpy is migrated to asynchronization memcpy, assuming in original code the source host memory is pageable memory. If this assumption doesn't hold, please append '.wait()' after memcpy to synchronize.")
+DEF_WARNING(CALL_SYNCTHREADS_IN_COND, 1115, "Call %0 inside a conditional branch statement may lead to undefined behavior")
```
Comment (zhimingwang36) [inline clang/lib/DPCT/Diagnostics.inc]: for 1115, use 1118 instead.

File: clang/test/dpct/syncthreads.cu [hunk 12]
```diff
@@ -185,6 +188,9 @@ __global__ void test10(float *arg1_ptr, int arg2_scalar, int arg3_scalar) {
       }
       // CHECK: var3 = var3 / 2;
       // CHECK-NEXT: /*
+      // CHECK-NEXT: DPCT1115:{{[0-9]+}}: Call __syncthreads inside a conditional branch statement may lead to undefined behavior
```
Comment (zhimingwang36) [inline clang/test/dpct/syncthreads.cu]: SYCL group functions and algorithms must be encountered in converged control flow. You should check this condition holds.

File: clang/lib/DPCT/SyncThreadsSideEffectsAnalyzer.cpp [hunk 13]
```diff
@@ -0,0 +1,169 @@
+//===---- SyncThreadsSideEffectAnalyzer.cpp ---------------------*- C++ -*-===//
```
Comment (zhimingwang36) [inline clang/lib/DPCT/SyncThreadsSideEffectsAnalyzer.cpp]: GroupFunctionAnalyzer.cpp 

add  a "//file purpose:  " section to desc the functionality.

File: clang/lib/DPCT/AnalysisInfo.h [hunk 14]
```diff
@@ -3766,6 +3770,14 @@ class DeviceFunctionInfo {
     return {};
   }
 
+  inline void setHasSideEffects(bool Val = true) { HasSideEffects = Val; }
+  inline bool hasSideEffects() const { return HasSideEffects; }
+
+  inline void setHasSideEffectsAnalyzed(bool Val = true) {
+    HasSideEffectsAnalyzed = Val;
```
Comment (zhimingwang36) [inline clang/lib/DPCT/AnalysisInfo.h]: Please add comment on; user of this flag for easy code reading.

File: clang/lib/DPCT/SyncThreadsSideEffectsAnalyzer.cpp (hunk 13 above)
Comment (yihwang-nv) [reply clang/lib/DPCT/SyncThreadsSideEffectsAnalyzer.cpp]: fixed

File: clang/lib/DPCT/AnalysisInfo.h (hunk 14 above)
Comment (yihwang-nv) [reply clang/lib/DPCT/AnalysisInfo.h]: fixed

File: clang/test/dpct/syncthreads.cu (hunk 12 above)
Comment (yihwang-nv) [reply clang/test/dpct/syncthreads.cu]: fixed

File: clang/lib/DPCT/Diagnostics.inc (hunk 11 above)
Comment (yihwang-nv) [reply clang/lib/DPCT/Diagnostics.inc]: fixed

**Review by zhimingwang36 — APPROVED** [review]
LGTM

**Pull request merged**